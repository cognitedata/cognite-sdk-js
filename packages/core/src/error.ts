// Copyright 2020 Cognite AS
import { X_REQUEST_ID } from './constants';
import type { HttpError } from './httpClient/httpError';
import { getHeaderField } from './httpClient/httpHeaders';

export class CogniteError extends Error {
  public status: number;
  public errorMessage: string;
  public requestId?: string;
  public missing?: object[];
  public duplicated?: unknown[];
  public extra?: unknown;
  /** @hidden */
  constructor(
    errorMessage: string,
    status: number,
    requestId?: string,
    otherFields: {
      missing?: object[];
      duplicated?: unknown[];
      extra?: unknown;
    } = {},
  ) {
    let message = `${errorMessage} | code: ${status}`;
    if (requestId) {
      message += ` | X-Request-ID: ${requestId}`;
    }
    const { missing, duplicated, extra } = otherFields;
    if (missing || duplicated) {
      message += `\n${JSON.stringify(otherFields, null, 2)}`;
    }
    super(message);
    Object.setPrototypeOf(this, CogniteError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript
    this.errorMessage = errorMessage;
    this.status = status;
    this.requestId = requestId;
    this.missing = missing;
    this.duplicated = duplicated;
    this.extra = extra;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }

  public toJSON() {
    const jsonObject: { [key: string]: unknown } = {
      status: this.status,
      message: this.errorMessage,
    };
    if (this.requestId) {
      jsonObject.requestId = this.requestId;
    }
    if (this.missing) {
      jsonObject.missing = this.missing;
    }
    if (this.duplicated) {
      jsonObject.duplicated = this.duplicated;
    }
    if (this.extra) {
      jsonObject.extra = this.extra;
    }
    return jsonObject;
  }
}

/** @hidden */
export function handleErrorResponse(err: HttpError) {
  let code: number;
  let duplicated: object[] | undefined;
  let extra: unknown;
  let message: string;
  let missing: object[] | undefined;
  let requestId: string | undefined;
  try {
    code = err.status;
    const data = err.data;
    duplicated = data.error.duplicated;
    message = data.error.message;
    missing = data.error.missing;
    extra = data.error.extra;
    requestId = getHeaderField(err.headers || {}, X_REQUEST_ID);
  } catch (_) {
    throw err;
  }
  throw new CogniteError(message, code, requestId, {
    duplicated,
    extra,
    missing,
  });
}
