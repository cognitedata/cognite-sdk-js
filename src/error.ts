// Copyright 2019 Cognite AS

import { AxiosError } from 'axios';

export class CogniteError extends Error {
  public status: number;
  public requestId?: string;
  public missing?: object[];
  public duplicated?: any[];
  /** @hidden */
  constructor(
    errorMessage: string,
    status: number,
    requestId?: string,
    extra: any = {}
  ) {
    let message = `${errorMessage} | code: ${status}`;
    if (requestId) {
      message += ` | X-Request-ID: ${requestId}`;
    }
    const { missing, duplicated } = extra;
    if (missing || duplicated) {
      message += `\n${JSON.stringify(extra, null, 2)}`;
    }
    super(message);
    Object.setPrototypeOf(this, CogniteError.prototype);
    this.status = status;
    this.requestId = requestId;
    this.missing = missing;
    this.duplicated = duplicated;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }
}

/** @hidden */
export function handleErrorResponse(err: AxiosError) {
  let code;
  let message;
  let requestId;
  let missing;
  let duplicated;
  try {
    code = err.response!.status;
    message = err.response!.data.error.message;
    missing = err.response!.data.error.missing;
    duplicated = err.response!.data.error.duplicated;
    requestId = (err.response!.headers || {})['X-Request-Id'];
  } catch (_) {
    throw err;
  }
  throw new CogniteError(message, code, requestId, { missing, duplicated });
}
