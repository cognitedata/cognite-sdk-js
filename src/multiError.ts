// Copyright 2019 Cognite AS

import { CogniteError } from './error';
import { HttpResponse } from './utils/http/basicHttpClient';

/** @hidden */
export interface MultiErrorRawSummary<RequestType, ResponseType> {
  succeded: RequestType[][];
  responses: HttpResponse<ResponseType>[];
  failed: RequestType[][];
  errors: (Error | CogniteError)[];
}

function serialiseErrors(errors: (Error | CogniteError)[]) {
  return errors.map(err => {
    if (err instanceof CogniteError) {
      return err;
    }
    const { message, name, stack } = err;
    return { name, message, stack };
  });
}

export class CogniteMultiError<RequestType, ResponseType> extends Error {
  public failed: RequestType[];
  public succeded: RequestType[];

  public status?: number;
  public requestId?: string;
  public missing: object[] = [];
  public duplicated: any[] = [];
  public statuses: number[] = [];
  public requestIds: string[] = [];
  public responses: ResponseType[] = [];
  public errors: (Error | CogniteError)[];

  constructor({
    succeded,
    responses,
    failed,
    errors,
  }: MultiErrorRawSummary<RequestType, ResponseType>) {
    super(
      `The API Failed to process some items\n${JSON.stringify(
        {
          succeded,
          responses,
          failed,
          errors: serialiseErrors(errors),
        },
        null,
        2
      )}`
    );
    Object.setPrototypeOf(this, CogniteMultiError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript

    this.responses = responses.map(response => response.data);
    this.errors = errors;
    this.succeded = succeded.reduce((all, current) => all.concat(current), []);
    this.failed = failed.reduce((all, current) => all.concat(current), []);
    this.status = (errors[0] as CogniteError).status;
    this.requestId = (errors[0] as CogniteError).requestId;

    for (const err of errors) {
      if (err instanceof CogniteError) {
        if (err.missing) {
          this.missing = this.missing.concat(err.missing);
        }
        if (err.duplicated) {
          this.duplicated = this.duplicated.concat(err.duplicated);
        }
        if (err.requestId) {
          this.requestIds.push(err.requestId);
        }
        this.statuses.push(err.status);
      }
    }

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(this.message).stack;
    }
  }
}
