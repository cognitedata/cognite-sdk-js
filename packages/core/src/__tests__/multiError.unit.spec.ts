import { CogniteMultiError } from '../multiError';
import { CogniteError } from '../error';

describe('Cognite multi error', () => {
  test('create with 2 fails and 1 success', () => {
    const errMsg = 'createAssets.arg0.items: size must be between 1 and 1000';
    const nestedErr = new CogniteError(errMsg, 400, 'r1', {
      missing: ['something'],
    });
    const nestedErr2 = new CogniteError(errMsg, 500, 'r2', {
      missing: ['more'],
      duplicated: ['this one'],
    });
    const err = new CogniteMultiError({
      succeded: [[2]],
      failed: [[0, 1]],
      errors: [nestedErr, nestedErr2],
      responses: [],
    });

    expect(err.succeded).toEqual([2]);
    expect(err.failed).toEqual([0, 1]);
    expect(err.statuses).toEqual([400, 500]);
    expect(err.status).toEqual(400);
    expect(err.requestId).toEqual('r1');
    expect(err.errors).toEqual([nestedErr, nestedErr2]);
    expect(err.message).toMatchSnapshot();
    expect(err.missing).toEqual(['something', 'more']);
    expect(err.duplicated).toEqual(['this one']);
    expect(err.requestIds).toEqual(['r1', 'r2']);
  });

  test('multierror serialises non-api errors', () => {
    const unknownError = new Error('unknown');

    const err = new CogniteMultiError({
      failed: [],
      errors: [unknownError],
      succeded: [],
      responses: [],
    });

    expect(err.message).toContain(`"message": "${unknownError.message}"`);
  });
});
