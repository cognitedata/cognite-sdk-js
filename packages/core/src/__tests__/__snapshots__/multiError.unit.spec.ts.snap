// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cognite multi error create with 2 fails and 1 success 1`] = `
"The API Failed to process some items
{
  \\"succeded\\": [
    [
      2
    ]
  ],
  \\"responses\\": [],
  \\"failed\\": [
    [
      0,
      1
    ]
  ],
  \\"errors\\": [
    {
      \\"status\\": 400,
      \\"message\\": \\"createAssets.arg0.items: size must be between 1 and 1000\\",
      \\"requestId\\": \\"r1\\",
      \\"missing\\": [
        \\"something\\"
      ]
    },
    {
      \\"status\\": 500,
      \\"message\\": \\"createAssets.arg0.items: size must be between 1 and 1000\\",
      \\"requestId\\": \\"r2\\",
      \\"missing\\": [
        \\"more\\"
      ],
      \\"duplicated\\": [
        \\"this one\\"
      ]
    }
  ]
}"
`;
